### PR TITLE
fix: fail fast on cyclic UDP references

### LINKED ISSUE #375: Fail fast on cyclic UDP references instead of relying on Python's recursion limit
## Context

Follow-up from review of #370 (resolve user-defined process references in process graphs).

A UDP referencing itself directly can't be stored — `upsert_udp` rejects it, since the UDP doesn't exist yet at the moment its own existence is checked (chicken-and-egg). But a real cycle *can* be constructed via a second `PUT`: store `A` as something valid, store `B` referencing `A`, then re-`PUT` `A` to reference `B` instead. `A` and `B` individually pass the existence check at each step, since the referenced id exists at that moment.

## Current behavior

Using a cyclic UDP (e.g. via `POST /services` or `POST /result`) fails cleanly today — a 4xx, not a 500 or a hang — and `resolve_udp_references`'s catch-all logs a warning (`exc_info=True`) before returning the graph unresolved. So this is not a correctness bug.

It does take noticeably longer than a normal request (~1.4s in local testing), which strongly suggests the failure path is Python's recursion limit (`RecursionError`) being hit and caught, rather than a fast, explicit check. `_fill_in_processes` (in the vendored `openeo_pg_parser_networkx` resolver) has no cycle guard of its own — the outer gate only checks the `predefined` namespace, so re-encountering the same `process_id` inside its own expansion re-triggers a deepcopy-and-recurse every time.

## Proposed enhancement

Track visited `(namespace, process_id)` pairs across a single resolution call (e.g. in `resolve_udp_references` / `_resolve` in `titiler/openeo/udp_resolution.py`) and raise a clear, immediate error the moment a cycle is detected, instead of relying on hitting the interpreter's recursion limit. This would:

- Fail in roughly constant time instead of burning a full recursion stack's worth of work and repeated `deepcopy` calls per request.
- Produce a specific, actionable error message (e.g. naming the cycle: `a -> b -> a`) instead of a generic "not found in registry" surfaced after the resolver gives up.

## Scope note

This is scoped to a user's own UDPs referencing each other — it can't be used to affect other users' requests, and today's behavior is a clean 4xx either way. Filed as a hardening/UX enhancement, not a bug.

### PR DESCRIPTION
**Pull request by arjxn-py**
## Description

Closes #375, follow-up from the review of #370.

UDPs can reference each other, and a user can make a loop out of that - one valid `PUT` at a time. Store `a`, store `b` pointing at `a`, then re-store `a` pointing at `b`. Neither write can be turned down on its own: the loop only exists after the second one lands.

Using such a graph already failed with a 4xx, so this is not a correctness fix. But it failed the slow, unhelpful way. The vendored resolver has no cycle guard, so it kept deep-copying and recursing until Python's recursion limit stopped it - 986 frames deep - and the `RecursionError` came back out as "process not found in registry", naming neither the cycle nor either UDP in it.

So we now walk the reference graph ourselves before inlining anything, keeping a stack of the UDPs being expanded, and raise the moment one is re-entered:

```
User-defined processes reference each other in a cycle: cycle_a -> cycle_b -> cycle_a
```

Measured on a two-UDP cycle:

```
before: 0.038s, RecursionError, graph handed back unresolved
after:  0.0001s, cycle named, 2 store lookups
```

## Testing

- [x] I have run the existing test suite and all tests pass
- [x] I have tested this change manually

## Documentation

- [x] I have updated the documentation accordingly
- [ ] My changes do not require documentation updates

## Checklist

- [x] My PR title follows [conventional commit format](https://www.conventionalcommits.org/) (e.g., `feat: add new feature`, `fix: resolve issue`)
- [x] My changes generate no new warnings
- [x] I updated the Helm chart version if applicable

## Related Issues

Fixes #

### CODE REVIEW

File: titiler/openeo/udp_resolution.py [hunk 1]
```diff
@@ -78,27 +85,65 @@ def strip_falsy_result_keys(process_graph: Any) -> None:
...
+    for process_id in udp_references(process_graph, predefined):
+        if process_id in chain:
+            raise CyclicUdpReference(chain[chain.index(process_id) :] + [process_id])
+        if process_id in checked:
+            continue
+        udp = get_udp_spec(process_id, namespace)
+        body = udp.get("process_graph") if isinstance(udp, dict) else None
+        if isinstance(body, dict):
+            chain.append(process_id)
+            _check_for_cycles(body, predefined, get_udp_spec, namespace, chain, checked)
```
Comment (emmanuelmathot) [inline titiler/openeo/udp_resolution.py:144]: _check_for_cycles has no depth limit. A long UDP chain (about 16000+ hops, no cycle) raises RecursionError. The blanket `except Exception` in resolve_udp_references (line 241) catches it and returns the graph unresolved. This causes the same confusing "not found in registry" error this PR aims to fix, just from chain depth instead of a cycle. Verified with a 20000-hop test chain: `result is graph` is True and the UDP process_id stays unresolved.

An iterative walk (or a networkx-based one, since networkx is already used in this package, see reader_requirements.py) would remove this limit.

File: titiler/openeo/udp_resolution.py [hunk 2]
```diff
@@ -174,7 +229,15 @@ def fetch_udp_spec(process_id: str, udp_namespace: str) -> Optional[dict]:
     graph = deepcopy(process_graph)
     strip_falsy_result_keys(graph)
     try:
+        # Before inlining anything: a cycle would otherwise be "found" only by
+        # exhausting the recursion limit, after a full stack's worth of
+        # deepcopying, and reported as a missing process.
+        _check_for_cycles(graph, predefined, fetch_udp_spec, namespace, [], set())
         return _resolve(graph, scratch, fetch_udp_spec, namespace, predefined)
```
Comment (emmanuelmathot) [inline titiler/openeo/udp_resolution.py:236]: fetch_udp_spec caches each UDP spec on first read. The cycle check reads it first, then _resolve reuses the same cached copy instead of re-reading. If the store changes between these two reads, resolution uses the old copy. This needs a store update inside a narrow window, so it is a low-risk edge case, not a common bug. Worth a one-line note in the docstring if this trade-off is intentional.

File: tests/test_udp_references.py [hunk 3]
```diff
@@ -604,6 +604,82 @@ def boom(*args, **kwargs):
...
+    )
+    resp = _store_udp(
+        client,
+        "cycle_a",
+        {"u1": {"process_id": "cycle_b", "arguments": {}, "result": True}},
+    )
+    assert resp.status_code in (200, 201), resp.text
+
+
+def test_cyclic_udp_reference_fails_with_the_cycle_named(app_with_auth):
```
Comment (emmanuelmathot) [inline tests/test_udp_references.py:628]: The new cycle tests cover POST /services and POST /validation only. PUT /process_graphs/{id}, PATCH /services/{id}, and POST /result also call _resolve_udp_references but have no cycle test. Adding tests for these three routes would stop a future change from silently breaking cycle handling there.

File: titiler/openeo/factory.py [hunk 4]
```diff
@@ -806,66 +866,22 @@ def validate_process_graph(
...
             # (a graph that extends a stored UDP is valid).
-            raw_pg = self._resolve_udp_references(raw_pg, user)
+            try:
+                raw_pg = self._resolve_udp_references(raw_pg, user)
+            except CyclicUdpReference as err:
+                # This endpoint reports what's wrong with a graph in its `errors`
+                # list, not as an HTTP error -- a cycle is a finding like any
+                # other, and there is no resolved graph left to validate.
+                return {
+                    "errors": [{"code": "ProcessGraphInvalid", "message": err.message}]
```
Comment (emmanuelmathot) [inline titiler/openeo/factory.py:880]: Minor: this uses the literal code "ProcessGraphInvalid" instead of err.code ("InvalidProcessGraph"). It matches the other error codes already used in this function, so it is not a new problem from this PR, just a pre-existing mismatch worth a follow-up.

**Review by emmanuelmathot — CHANGES_REQUESTED** [review]
A few findings from review. Most important: a recursion depth issue in the new cycle check.

Two more notes that fall outside the diff, so they are not attached to a line:
- factory.py PATCH /services/{service_id} (around line 1289): this call can now raise CyclicUdpReference (HTTP 422), but the route's `responses` dict (around line 1231) does not list 422. POST /services documents 422 for the same case.
- factory.py POST /result (around line 1450): same issue. The `responses` dict (around line 1425) lists only 200.
